src/server.c

```diff
@@ -693,34 +693,23 @@ dictType clientDictType = {
     .no_value = 1               /* no values in this dict */
 };
 
-int htNeedsShrink(dict *dict) {
-    long long size, used;
-
-    size = dictBuckets(dict);
-    used = dictSize(dict);
-    return (size > DICT_HT_INITIAL_SIZE &&
-            (used * HASHTABLE_MIN_FILL <= size));
-}
-
 /* In cluster-enabled setup, this method traverses through all main/expires dictionaries (CLUSTER_SLOTS)
  * and triggers a resize if the percentage of used buckets in the HT reaches (100 / HASHTABLE_MIN_FILL)
- * we resize the hash table to save memory.
+ * we shrink the hash table to save memory, or expand the hash when the percentage of used buckets reached
+ * 100.
  *
  * In non cluster-enabled setup, it resize main/expires dictionary based on the same condition described above. */
 void tryResizeHashTables(int dbid) {
     redisDb *db = &server.db[dbid];
+    int dicts_per_call = min(CRON_DICTS_PER_DB, db->dict_count);
     for (dbKeyType subdict = DB_MAIN; subdict <= DB_EXPIRES; subdict++) {
-        if (dbSize(db, subdict) == 0) continue;
-
-        if (db->sub_dict[subdict].resize_cursor == -1)
-            db->sub_dict[subdict].resize_cursor = findSlotByKeyIndex(db, 1, subdict);
-
-        for (int i = 0; i < CRON_DBS_PER_CALL && db->sub_dict[subdict].resize_cursor != -1; i++) {
+        for (int i = 0; i < dicts_per_call; i++) {
             int slot = db->sub_dict[subdict].resize_cursor;
             dict *d = (subdict == DB_MAIN ? db->dict[slot] : db->expires[slot]);
-            if (htNeedsShrink(d))
-                dictShrinkToFit(d);
-            db->sub_dict[subdict].resize_cursor = dbGetNextNonEmptySlot(db, slot, subdict);
+            if (dictShrinkIfNeeded(d) == DICT_ERR) {
```

Comment:
do we really need the return value (and the other changes it brought)?
seems to me we can just call these two functions one after the other.

Reply:
If we shrink the dict at first, then there is no need to go through those ifs to tell whether we can expand. This saves some unnecessary cost.